Null http route hosts will cause panic

When I apply a ingress.yaml like this
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: test-ingress3
spec:
  rules:
  - host: test1.test.com
  - host: test2.test.com
    http:
      paths:
      - backend:
          serviceName: network-test
          servicePort: 9001
        path: /
```

The error log is 
```
E0808 09:58:00.092373       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/go/src/github.com/heptio/contour/internal/dag/builder.go:315
/go/src/github.com/heptio/contour/internal/dag/builder.go:131
/go/src/github.com/heptio/contour/internal/contour/cachehandler.go:45
/go/src/github.com/heptio/contour/internal/contour/holdoff.go:56
/go/src/github.com/heptio/contour/internal/contour/k8s.go:90
/go/src/github.com/heptio/contour/internal/contour/k8s.go:59
/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:554
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203
/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546
/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:390
/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71
/usr/local/go/src/runtime/asm_amd64.s:2361
time="2018-08-08T09:58:00Z" level=info msg="forcing update" context=HoldoffNotifier last update=2562047h47m16.854775807s
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xdae82f]

goroutine 81 [running]:
github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x12ad8c0, 0x1e32930)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/heptio/contour/internal/dag.(*Builder).compute(0xc4200e8b00, 0x14eb188)
	/go/src/github.com/heptio/contour/internal/dag/builder.go:315 +0x178f
github.com/heptio/contour/internal/dag.(*Builder).Compute(0xc4200e8b00, 0x0)
	/go/src/github.com/heptio/contour/internal/dag/builder.go:131 +0x77
github.com/heptio/contour/internal/contour.(*CacheHandler).OnChange(0xc4201da7e0, 0xc4200e8b00)
	/go/src/github.com/heptio/contour/internal/contour/cachehandler.go:45 +0x95
github.com/heptio/contour/internal/contour.(*HoldoffNotifier).OnChange(0xc4200c6870, 0xc4200e8b00)
	/go/src/github.com/heptio/contour/internal/contour/holdoff.go:56 +0x1b5
github.com/heptio/contour/internal/contour.(*ResourceEventHandler).update(0xc4200e8af0)
	/go/src/github.com/heptio/contour/internal/contour/k8s.go:90 +0x3d
github.com/heptio/contour/internal/contour.(*ResourceEventHandler).OnAdd(0xc4200e8af0, 0x143e1c0, 0xc4202d0c60)
	/go/src/github.com/heptio/contour/internal/contour/k8s.go:59 +0x188
github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x429c80, 0xc42006b5a0, 0xc42006b5a0)
	/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:554 +0x21a
github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0xc420531df0, 0x429799, 0xc4203180d0)
	/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:203 +0x9c
github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548 +0x81
github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc42006b768)
	/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc420531f68, 0xdf8475800, 0x0, 0x1282f01, 0xc42040c000)
	/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc42006b768, 0xdf8475800, 0xc42040c000)
	/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc42040e400)
	/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546 +0x78
github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache.(*processorListener).(github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache.run)-fm()
	/go/src/github.com/heptio/contour/vendor/k8s.io/client-go/tools/cache/shared_informer.go:390 +0x2a
github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc4203f9800, 0xc4201f0140)
	/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/src/github.com/heptio/contour/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```

I think this grammar maybe not effective and correct， but Kubernetes accept it, so I hope contour return a error instead of  do  panic.
Thanks

--- a/internal/dag/builder.go
+++ b/internal/dag/builder.go
@@ -312,6 +312,11 @@ func (b *Builder) compute() *DAG {
 			if host == "" {
 				host = "*"
 			}
+			if rule.IngressRuleValue.HTTP == nil {
+				// HTTP can be nil as rule.IngressRuleValue's zero value contains
+				// a nil HTTP key.
+				continue
+			}
 			for n := range rule.IngressRuleValue.HTTP.Paths {
 				path := rule.IngressRuleValue.HTTP.Paths[n].Path
 				if path == "" {

--- a/internal/dag/builder_test.go
+++ b/internal/dag/builder_test.go
@@ -64,6 +64,21 @@ func TestDAGInsert(t *testing.T) {
 			}},
 		},
 	}
+
+	// i2a is missing a http key from the spec.rule.
+	// see issue 606
+	i2a := &v1beta1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kuard",
+			Namespace: "default",
+		},
+		Spec: v1beta1.IngressSpec{
+			Rules: []v1beta1.IngressRule{{
+				Host: "test1.test.com",
+			}},
+		},
+	}
+
 	// i3 is similar to i2 but includes a hostname on the ingress rule
 	i3 := &v1beta1.Ingress{
 		ObjectMeta: metav1.ObjectMeta{
@@ -914,6 +929,12 @@ func TestDAGInsert(t *testing.T) {
 					),
 				},
 			},
+		},
+		"insert ingress with missing spec.rule.http key": {
+			objs: []interface{}{
+				i2a,
+			},
+			want: []Vertex{},
 		},
 		"insert ingress w/ host name and single backend": {
 			objs: []interface{}{

--- a/internal/e2e/rds_test.go
+++ b/internal/e2e/rds_test.go
@@ -1495,6 +1495,80 @@ func TestRDSAssertNoDataRaceDuringInsertAndStream(t *testing.T) {
 	}
 }
 
+// issue 606: spec.rules.host without a http key causes panic.
+// apiVersion: extensions/v1beta1
+// kind: Ingress
+// metadata:
+//   name: test-ingress3
+// spec:
+//   rules:
+//   - host: test1.test.com
+//   - host: test2.test.com
+//     http:
+//       paths:
+//       - backend:
+//           serviceName: network-test
+//           servicePort: 9001
+//         path: /
+//
+// note: this test caused a panic in dag.Builder, but testing the
+// context of RDS is a good place to start.
+func TestRDSIngressSpecMissingHTTPKey(t *testing.T) {
+	rh, cc, done := setup(t)
+	defer done()
+
+	i1 := &v1beta1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-ingress3",
+			Namespace: "default",
+		},
+		Spec: v1beta1.IngressSpec{
+			Rules: []v1beta1.IngressRule{{
+				Host: "test1.test.com",
+			}, {
+				Host: "test2.test.com",
+				IngressRuleValue: v1beta1.IngressRuleValue{
+					HTTP: &v1beta1.HTTPIngressRuleValue{
+						Paths: []v1beta1.HTTPIngressPath{{
+							Path: "/",
+							Backend: v1beta1.IngressBackend{
+								ServiceName: "network-test",
+								ServicePort: intstr.FromInt(9001),
+							},
+						}},
+					},
+				},
+			}},
+		},
+	}
+	rh.OnAdd(i1)
+
+	s1 := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "network-test",
+			Namespace: "default",
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{{
+				Name:       "http",
+				Protocol:   "TCP",
+				Port:       9001,
+				TargetPort: intstr.FromInt(8080),
+			}},
+		},
+	}
+	rh.OnAdd(s1)
+
+	assertRDS(t, cc, []route.VirtualHost{{
+		Name:    "test2.test.com",
+		Domains: []string{"test2.test.com", "test2.test.com:80"},
+		Routes: []route.Route{{
+			Match:  prefixmatch("/"), // match all
+			Action: routecluster("default/network-test/9001"),
+		}},
+	}}, nil)
+}
+
 func assertRDS(t *testing.T, cc *grpc.ClientConn, ingress_http, ingress_https []route.VirtualHost) {
 	t.Helper()
 	assertEqual(t, &v2.DiscoveryResponse{

